smf/clickhouse/method_script.sh

```diff
@@ -28,13 +28,78 @@ ipadm show-addr "$DATALINK/ll" || ipadm create-addr -t -T addrconf "$DATALINK/ll
 ipadm show-addr "$DATALINK/omicron6"  || ipadm create-addr -t -T static -a "$LISTEN_ADDR" "$DATALINK/omicron6"
 route get -inet6 default -inet6 "$GATEWAY" || route add -inet6 default -inet6 "$GATEWAY"
 
-args=(
-  "--log-file" "/var/tmp/clickhouse-server.log"
-  "--errorlog-file" "/var/tmp/clickhouse-server.errlog"
-  "--"
-  "--path" "${DATASTORE}"
-  "--listen_host" "$LISTEN_ADDR"
-  "--http_port" "$LISTEN_PORT"
-)
-
-exec /opt/oxide/clickhouse/clickhouse server "${args[@]}" &
+# Retrieve addresses of the other clickhouse nodes, order them and assign them to be a replica node.
```

Comment:
Something important to consider:

We already have a rack, in the field, which we are not able to update via RSS. We can **update** it, but we cannot delete all data from scratch.

This rack already has a clickhouse dataset provisioned to it, using the "old mechanism".

If we update that rack in-place, there will be:
- No keepers
- Still the single provisioned clickhouse node
- ... which will try to use this new method script

I believe if we do that, the old clickhouse zone will fail to boot.

Reply:
One possible path forward, that I *think* is straightforward, but may require some mechanical work:

## Before-this-PR: One service type

- The service `clickhouse` implies "single-node-clickhouse"
- No keepers exist

## This PR, as is: Two service types

- We have a "service type" for `clickhouse` and for `clickhouse-keeper`
- It is implied that `clickhouse` is a replica, acting within a multi-replica system, managed by keepers
- This may prove problematic on a system where "single node clickhouse" already exists

## Proposal: Three different service types

- `clickhouse` can, as it did before this PR, imply "single-node-clickhouse"
- `clickhouse-keeper` stays as-is -- it implies the keeper side of a multi-node cluster
- `clickhouse-replica` is a new service type, which represents a replica acting in a multi-node system

With this proposal, we can provision keepers + replicas on new systems and test them, but we'll keep all the code to launch a "single-node-clickhouse" as it was before this PR, so the use-case for our existing customers can stay the same.

**Later on**, when Nexus provisions services, it can ensure there are sufficient keeper / replica nodes, and tear down the single-node clickhouses.

Reply:
oh no! I thought we still had one more go at deleting data :(

hmmm, let me think about this one. Having three services can be a way forward, but it'd be really nice to be able update the current use-case to have a replicated set up.

Are we allowed to run one off jobs? That way we could run a migration just the once :thinking:

Reply:
If you want to run "one-off jobs", you'd need to define what that means.

## "The right way"

I think, in the limit, the "right way" to re-provision would be using a reliable persistent workflow in Nexus to ensure that necessary services are provisioned -- but this code would need to deal with the fact that "there are many instances of Nexus running at once, that need to be able to converge on the same result".

This is more complication than just "do an allocation once on Nexus boot", because:
- There many Nexus instances running, and they may be running concurrently
- These Nexus instances may crash arbitrarily at any point, and should still converge on the same result
  - This implies that service provisioning must be idempotent

This is doable (see: https://github.com/oxidecomputer/omicron/issues/732 -- I need to plumb generation number info through the sled agent API first, to make them idempotent, before making these calls)

## ... is that doable in a few days?

Honestly, I don't think so. It's a big chunk of work. Worth doing! But not trivial.

Reply:
oof, yeah, ok I get it now. I was hoping we could run a migration to replicated databases _before_ the update, but it seems like this would be much more hacky/error prone/unprofessional that I thought.

Right, I'll follow the 3 service approach, I'll look into it and may ping you if I get stuck :smile: